enhancements/automatic-catalog-switching-annotations.md

```diff
@@ -0,0 +1,343 @@
+  - example resolved value: 1
+- {ocp_patch_version}
+  - example resolved value: 1
+- {ocp_update_channels}
+
+#### Distribution version
+
+Declaring generic distribution variables leaves gaps in support since there
+are many distributions to consider and could lead to confusion when a 
+particular distribution is not supported.
+
+- {distro_major_version}
+  - example resolved value: 1
+- {distro_minor_version}
+  - example resolved value: 1
+- {distro_patch_version}
+  - example resolved value: 1
+
+### CatalogSource updates via annotations
+
+The mechanism for declaring an image using templates involves an annotation called
+`olm.catalogImageTemplate` whose value consists of the image reference with one or more templates included.
+A controller (which can exist in either the OLM catalog operator or within a standalone operator) is 
+responsible for watching `CatalogSource` instances looking for the `olm.catalogImageTemplate` annotation.
+Upon identifying a candidate `CatalogSource` the template values can be resolved and assuming all values
+are available, the `spec.image` of the `CatalogSource` is updated with the resolved value ONLY if the
+resolved value and the `spec.image` values differ. Once the `spec.image` is updated the OLM catalog 
+operator will handle the image update as if a user had changed the image value manually. 
+
+Consider the following example, where a catalog can be switched out based on the Kubernetes version in use:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: dynamic-catalog
+  namespace: olm
+  annotations:
+    olm.catalogImageTemplate: "quay.io/kube-release-v{kube_major_version}/catalog:v{kube_major_version}.{kube_minor_version}"
+spec:
+  sourceType: grpc
+  image: quay.io/kube-release-v1/catalog:v1.20
+  displayName: My always up2date Catalog
+  publisher: Red Hat
+```
+
+The image reference would resolve to `quay.io/kube-release-v1/catalog:v1.19` for a cluster running Kubernetes
+1.19 and would render `quay.io/kube-release-v1/catalog:v1.20` for Kubernetes 1.20. 
+
+The Kubernetes version can be obtained from one of these sources:
+
+- using Clientset from k8s.io/client-go/kubernetes
+
+   ```
+   serverVersion, err := clientset.Discovery().ServerVersion()
+   ```
+
+- the api server `/version` endpoint using the `gitVersion` field (e.g. v1.17.1+6af3663)
+
+Consider another example which uses an arbitrary kubernetes manifest as a source for template substitution:
+
+```yaml
+apiVersion: foo.example.com/v1
+kind: Sample
+metadata:
+  name: MySample
+  namespace: ns
+spec:
+  foo:
+    bar: v2
+```
+
+The catalog source references the `Sample` manifest above like this:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: dynamic-catalog
+  namespace: olm
+  annotations:
+    olm.catalogImageTemplate: "quay.io/sample/catalog:{group:foo.example.com,version:v1,kind:Sample,name:MySample,namespace:ns,jsonpath:spec.foo.bar}"
```

Comment:
Do we see any reason to optimize for referencing variables from THIS CatalogSource?  Like the downward API?

If we wanted to support that, then you could dispense with group, version, kind, name, namespace and only have jsonpath.  i.e. jsonpath by itself, references THIS resource.

Reply:
I like the idea, but I don't think there's anything in the catalog source spec to use as a template value that would be of interest. I suppose you could reference information under metadata or status. This could be something we add later if we find a need.